python/python/bystro/supervised_ppca/gf_generative_pt.py

```diff
@@ -0,0 +1,715 @@
+        """
+        pd = self.prior_options
+
+        def log_prior(trainable_variables):
+            list_gamma_log_probs = []
+            for k in self.n_groups:
+                list_gamma_log_probs.append(
+                    Gamma(pd["alpha"], pd["beta"]).log_prob(
+                        trainable_variables[k + 1]
+                    )
+                )
+
+            gamma_log_probs = torch.stack(list_gamma_log_probs)
+            return torch.mean(gamma_log_probs)
+
+        return log_prior
+
+    def _fill_prior_options(self, prior_options):
+        """
+        Fills in options for prior parameters
+
+        Paramters
+        ---------
+        new_dict : dictionary
+            The prior parameters used to specify the prior
+        """
+        default_dict = {"alpha": 3.0, "beta": 3.0}
+        new_dict = {**default_dict, **prior_options}
+        return new_dict
+
+    def _initialize_variables(self, X):
+        """
+        Initializes the variables of the model
+
+        Parameters
+        ----------
+        X : np.array-like,(n_samples,p)
+            The data
+
+        Returns
+        -------
+        W_ : tf.Variable-like,(n_components,p)
+            The loadings of our latent factor model
+
+        sigmal_ : list
+            A list of the isotropic noises for each group
+        """
+        model = dp.PCA(self.n_components)
+        S_hat = model.fit_transform(X)
+        W_init = model.components_.astype(np.float32)
+        W_ = torch.tensor(W_init, requires_grad=True)
+        X_recon = np.dot(S_hat, W_init)
+        diff = np.mean((X - X_recon) ** 2)
+        sinv = softplus_inverse_np(diff * np.ones(1).astype(np.float32))
+        sigmal_ = [
+            torch.tensor(sinv, requires_grad=True) for i in range(self.n_groups)
+        ]
+        return W_, sigmal_
+
+    def _store_instance_variables(self, trainable_variables):
+        """
+        Saves the learned variables
+
+        Parameters
+        ----------
+        trainable_variables : list
+            List of variables learned by pytorch
+
+        Sets
+        ----
+        W_ : np.array-like,(n_components,p)
+            The loadings
+
+        sigmas_ : np.array-like,(p,)
+            The diagonal variances
+        """
+        self.W_ = trainable_variables[0].detach().numpy()
+        self.sigmas_ = np.zeros(self.p)
+        for k in range(self.n_groups):
+            sigma2 = nn.Softplus()(trainable_variables[k + 1])
+            self.sigmas_[self.groups == k] = sigma2
+
+    def _test_inputs(self, X, groups):
+        """
+        Just tests to make sure data is numpy array
+        """
+        if isinstance(X, np.ndarray) is False:
+            raise ValueError("X is numpy array")
+        if isinstance(groups, np.ndarray) is False:
+            raise ValueError("groups is numpy array")
+        if X.shape[1] != len(groups):
+            raise ValueError("Dimensions do not match")
+
+
+class FactorAnalysispt(BasePCASGDModel):
```

Comment:
I think `PT` wants to be capitalized here?

Reply:
Do we? Genuine question.

Reply:
PEP8 recommends CapWords style for class names.  Would recommend `FactorAnalysisPT` or `FactorAnalysisPt`, but not `FactorAnalysispt`, which reads as "Factor Analysispt" and is difficult to interpret.

In general, it's better to name classes according to their interface (i.e. what they do) rather than their implementation (how they do it), so if we could get away without mentioning pytorch in the class name at all I'd recommend that, especially since we're converting everything back to numpy arrays after training anyway.   But it's better to have consistent class names within a module so we should probably do that either everywhere or not at all.

Reply:
Yeah now that I have separate files it doesn't make sense. I just created the distinction when I was using both tensorflow and pytorch. Will eliminate.